src/berryimu.py

```diff
@@ -0,0 +1,243 @@
+        # This loop must run for the same amount of time each loop to accurately complete the complementary filter. 
+        while True:
+            now = time.time() 
+            self.velocity = self.read_angular_velocity()
+            self.compass_heading = self.calculate_compass_heading()
+            self.acceleration = self.get_acceleration()
+            self.orientation = self.calculate_orientation()
+            elapsed = time.time() - now
+            time.sleep(dt - elapsed) # full iteration takes 80 ms
+
+    def calculate_compass_heading(self):
+        raw_data = self.i2cbus.read_i2c_block_data(constants.MAG_ADDRESS,  constants.OUT_X_L, 6)
+        values = utils.parse_output(raw_data)
+        if self.calibrate:
+            # hard iron distortion correction
+            values.x -= (self.hard_iron_x_min + self.hard_iron_x_max) /2
+            values.y -= (self.hard_iron_y_min + self.hard_iron_y_max) /2
+
+            # #soft iron distortion correction 
+            scale_y = self.soft_iron_y_max - self.soft_iron_y_min
+            scale_x = self.soft_iron_x_max - self.soft_iron_x_min 
+            ratio = scale_y / scale_x
+            values.x *= ratio
+            values.y *= ratio
+        
+        # tilt compensation.
+        values.x = values.x * math.cos(self.euler_angles['pitch']) + values.z * math.sin(self.euler_angles['pitch'])
+
+        values.y = values.x * math.sin(self.euler_angles['roll'])  * math.sin(self.euler_angles['pitch']) + values.y * math.cos(self.euler_angles['roll']) 
+        - values.z * math.sin(self.euler_angles['roll']) * math.cos(self.euler_angles['pitch'])
+
+        heading = 180 * math.atan2(values.y, values.x) / math.pi
+        if heading < 0:
+              heading += 360
+
+        return heading
+        
+    def get_acceleration(self):
+       raw_data = self.i2cbus.read_i2c_block_data(constants.AG_ADDRESS, constants.OUTX_L_XL, 6)
+       self.accelerometer_raw = utils.parse_output(raw_data)
+      
+       #convert the raw readings to acceleration in m/s^2
+       self.acceleration.x =  self.accelerometer_raw.x * constants.A_GAIN / 1000 * 9.8065
+       self.acceleration.y =  self.accelerometer_raw.y * constants.A_GAIN / 1000 * 9.8065
+       self.acceleration.z =  self.accelerometer_raw.z * constants.A_GAIN / 1000 * 9.8065
+
+       return self.acceleration
+    
+    def read_angular_velocity(self):
+       raw_data = self.i2cbus.read_i2c_block_data(constants.AG_ADDRESS, constants.OUTX_L_G, 6)
+       values = utils.parse_output(raw_data)
+
+       # convert raw readings to velocity in degrees/s
+       self.velocity.x = values.x *  constants.G_GAIN
+       self.velocity.y = values.y *  constants.G_GAIN
+       self.velocity.z = values.z *  constants.G_GAIN
+
+       return self.velocity
+
+    def get_gyro_angles(self):
+         # multiply the velocity by the time between readings to get the angle moved. 
+        self.gyro_angles.x = math.radians(self.velocity.x * dt)
```

Comment:
where is dt defined?

Reply:
top of this file, can also move to constants if thats better.